jerry-core/ecma/builtin-objects/ecma-builtin-helpers-sort.c

```diff
@@ -17,133 +17,137 @@
 #include "ecma-globals.h"
 
 /**
- * Function used to reconstruct the ordered binary tree.
- * Shifts 'index' down in the tree until it is in the correct position.
+ * Function used to merge two arrays for merge sort.
+ * Arrays are stored as below:
+ * First  -> array_a_p [left_idx : right_idx - 1]
+ * Second -> array_a_p [right_idx : end_idx - 1]
+ * Output -> array_b_p
  *
  * @return ecma value
  *         Returned value must be freed with ecma_free_value.
  */
 static ecma_value_t
-ecma_builtin_helper_array_to_heap (ecma_value_t *array_p, /**< heap data array */
-                                   uint32_t index, /**< current item index */
-                                   uint32_t right, /**< right index is a maximum index */
-                                   ecma_value_t compare_func, /**< compare function */
-                                   const ecma_builtin_helper_sort_compare_fn_t sort_cb) /**< sorting cb */
+ecma_builtin_helper_array_merge_sort_bottom_up (ecma_value_t *array_a_p, /**< arrays to merge */
+                                                uint32_t left_idx, /**< first array begin */
+                                                uint32_t right_idx, /**< first array end */
+                                                uint32_t end_idx, /**< second array end */
+                                                ecma_value_t *array_b_p, /**< output array */
+                                                ecma_value_t compare_func, /**< compare function */
+                                                const ecma_builtin_helper_sort_compare_fn_t sort_cb) /**< sorting cb */
 {
-  /* Left child of the current index. */
-  uint32_t child = index * 2 + 1;
-  ecma_value_t swap = array_p[index];
-
   ecma_value_t ret_value = ECMA_VALUE_EMPTY;
+  uint32_t i = left_idx, j = right_idx;
 
-  while (child <= right)
+  for (uint32_t k = left_idx; k < end_idx; k++)
   {
-    if (child < right)
-    {
-      /* Compare the two child nodes. */
-      ecma_value_t child_compare_value = sort_cb (array_p[child], array_p[child + 1], compare_func);
+    ecma_value_t compare_value = ecma_make_number_value (ECMA_NUMBER_ZERO);
 
-      if (ECMA_IS_VALUE_ERROR (child_compare_value))
+    if (j < end_idx)
+    {
+      compare_value = sort_cb (array_a_p[i], array_a_p[j], compare_func);
```

Comment:
Shouldn't we also check that `i < right_idx`? It seems to me that otherwise we could end up comparing an element to itself, which I think should be avoided. (e.g. merging `1,3` with `5`, we would compare 1-5, 3-5, and then `i` will overflow the left array, and we compare 5-5, even though the result of the comparision will be unused)